packages/swing-store/src/transcriptStore.js

```diff
@@ -314,21 +329,124 @@ export function makeTranscriptStore(
+      // so we delete the transcript.${vatID}.current record, and add a
+      // .startPos one to replace it
+      noteExport(spanMetadataKey({ vatID, isCurrent: true }), undefined);
+      const newRec = spanRec(vatID, startPos, endPos, hash, false, incarnation);
+      noteExport(spanMetadataKey(newRec), JSON.stringify(newRec));
+    }
+  }
+
   /**
-   * Delete all transcript data for a given vat (for use when, e.g., a vat is terminated)
    *
    * @param {string} vatID
+   * @returns {boolean}
    */
-  function deleteVatTranscripts(vatID) {
+  function hasSpans(vatID) {
+    // note the LIMIT 1: we aren't really fetching all spans
+    const spans = sqlGetSomeVatSpans.all(vatID, 1);
+    return !!spans.length;
+  }
+
+  /**
+   * Delete at most 'budget' transcript spans, and their items.
+   *
+   * @param {string} vatID
+   * @param {number} budget
+   * @returns {{ done: boolean, cleanups: number }}
+   */
+  function deleteSomeVatTranscripts(vatID, budget) {
+    ensureTxn();
+    assert(budget >= 1);
+    let cleanups = 0;
+
+    // This query is ORDER BY startPos DESC, so we delete the newest
+    // spans first. If the kernel failed to call stopUsingTranscript,
+    // we might encounter an isCurrent=1 span, but we can delete those
+    // too.
+    const deletions = sqlGetSomeVatSpans.all(vatID, budget);
+
+    if (!deletions.length) {
+      return { done: true, cleanups };
+    }
+    for (const rec of deletions) {
+      // If rec.isCurrent is true, this will remove the
+      // transcript.$vatID.current export-data record. If false, it
+      // will remove the transcript.$vatID.$startPos record.
+      noteExport(spanMetadataKey(rec), undefined);
+      sqlDeleteVatSpan.run(vatID, rec.startPos);
+      sqlDeleteSomeItems.run(vatID, rec.startPos, rec.endPos);
+      cleanups += 1;
+    }
+    if (hasSpans(vatID)) {
+      return { done: false, cleanups };
+    }
+    return { done: true, cleanups };
+  }
+
+  function deleteAllVatTranscripts(vatID) {
     ensureTxn();
+    // we can't use .iterate here because noteExport writes to the DB,
+    // and we can't have overlapping queries
     const deletions = sqlGetVatSpans.all(vatID);
     for (const rec of deletions) {
       noteExport(spanMetadataKey(rec), undefined);
     }
+    // might need to delete the .current record, if the caller failed
+    // to call stopUsingTranscript()
     sqlDeleteVatItems.run(vatID);
     sqlDeleteVatSpans.run(vatID);
   }
 
+  /**
+   * Delete some or all transcript data for a given vat (for use when,
+   * e.g., a vat is terminated)
+   *
+   * @param {string} vatID
+   * @param {number} [budget]
+   * @returns {{ done: boolean, cleanups: number }}
+   */
+  function deleteVatTranscripts(vatID, budget = undefined) {
+    if (budget) {
+      return deleteSomeVatTranscripts(vatID, budget);
+    } else {
+      deleteAllVatTranscripts(vatID);
+      // no budget? no accounting.
+      return { done: true, cleanups: 0 };
+    }
+  }
```

Comment:
yep, good

Also, it's good to remove that `might need to delete the .current record, if the caller failed to call stopUsingTranscript()` comment from `deleteAllVatTranscripts`, because it's distracting: the `noteExport(spanMetadataKey(rec), undefined)` will already delete the .current record, if/when it encounters an `isCurrent = true` row. I think I was confused by the similar-but-different pattern in snapStore, where there's an *additional* metadata record.